File: src/tui/icons.ts
/**
 * Nerd Font glyph registry for the Ink TUI.
 *
 * A patched **Nerd Font** is a hard requirement for the TUI. Set
 * `WS_STATUS_GLYPHS=ascii` to fall back to plain markers if a terminal
 * renders the private-use area as tofu.
 *
 * Every glyph here occupies exactly **one** terminal column: Nerd Font icons
 * live in the private-use area (U+E000–U+F8FF), which `visibleWidth` counts as
 * width 1. Do not add codepoints from the emoji or CJK ranges — those are
 * double-width and break Ink's box layout. Plain-report badges
 * (`badgeForChange` in `changes.ts`) stay emoji; they are a separate surface.
 */

import { visibleWidth } from '../helpers.js';
import { getTheme } from './theme.js';
import type { Segment } from './theme.js';
import type { FileChange, SyncStatus } from '../types.js';
import type { FileStatusLetter } from './model/types.js';

/** True when the operator opted out of Nerd Font glyphs. */
export const asciiGlyphs = process.env.WS_STATUS_GLYPHS === 'ascii';

/** Pick the Nerd Font glyph unless ASCII fallback is active. */
function glyph(nerd: string, ascii: string): string {
  return asciiGlyphs ? ascii : nerd;
}

/* ── Structure ──────────────────────────────────────────────────────────── */

/** Fold chevrons — width 1 in both modes. */
export const FOLD_EXPANDED = '▾';
export const FOLD_COLLAPSED = '▸';

/** Cursor accent bar painted in the left-most tree column. */
export const CURSOR_BAR = '▌';

/** Vertical rule between panes and inside the diff gutter. */
export const RULE = '│';

export const ICON_WORKSPACE = glyph('', '#');
export const ICON_REPO = glyph('', '@');
/**
 * Linked `git worktree` checkout (plain-report `🔗`).
 * Nerd: nf-oct-link (``); ASCII: `L`.
 */
export const ICON_LINKED_WORKTREE = glyph('', 'L');
export const ICON_BRANCH = glyph('', '&');
export const ICON_FOLDER = glyph('', '/');
export const ICON_FOLDER_OPEN = glyph('', '/');
export const ICON_CLEAN = glyph('', '.');
export const ICON_IGNORED = glyph('', '~');
export const ICON_AHEAD = glyph('', '^');
export const ICON_BEHIND = glyph('', 'v');
export const ICON_DIVERGED = glyph('', 'Y');
export const ICON_NO_UPSTREAM = glyph('', '?');
export const ICON_SYNCED = glyph('', '=');
/**
 * Plain-report `✅` — HEAD is an ancestor of the default-branch tip.
 * Nerd: nf-fa-check-circle (``); ASCII: `M`.
 */
export const ICON_MERGED_INTO_DEFAULT = glyph('', 'M');
/**
 * Plain-report `🌱` — HEAD is not merged into default.
 * Nerd: nf-fa-tree (``); ASCII: `o`.
 */
export const ICON_OPEN_VS_DEFAULT = glyph('', 'o');
/**
 * Reviewed mark on a dirty file row (GitLens-style "viewed").
 * Nerd: nf-fa-eye (``); ASCII: `*`. Distinct from ICON_CLEAN / ICON_SYNCED.
 */
export const ICON_VIEWED = glyph('', '*');

/** Help-panel group icons. */
export const ICON_MOVE = glyph('', '+');
export const ICON_DIFF = glyph('', '%');
export const ICON_APP = glyph('', '!');

/** Font the TUI is designed against — surfaced in `--help` and the help panel. */
export const REQUIRED_FONT = 'MesloLGM Nerd Font Mono';
/* ── File type devicons ─────────────────────────────────────────────────── */

interface FileIcon {
  glyph: string;
  color: string;
}

const DEFAULT_FILE_GLYPH = '';

/** Extension → devicon. Codepoints match the nvim-web-devicons vocabulary. */
const EXTENSION_ICONS: Record<string, Omit<FileIcon, 'color'> & { color?: string }> = {
  ts: { glyph: '', color: '#519aba' },
  mts: { glyph: '', color: '#519aba' },
  cts: { glyph: '', color: '#519aba' },
  tsx: { glyph: '', color: '#519aba' },
  js: { glyph: '', color: '#cbcb41' },
  mjs: { glyph: '', color: '#cbcb41' },
  cjs: { glyph: '', color: '#cbcb41' },
  jsx: { glyph: '', color: '#519aba' },
  json: { glyph: '', color: '#cbcb41' },
  md: { glyph: '', color: '#519aba' },
  mdx: { glyph: '', color: '#519aba' },
  py: { glyph: '', color: '#ffbc03' },
  sh: { glyph: '', color: '#89e051' },
  bash: { glyph: '', color: '#89e051' },
  zsh: { glyph: '', color: '#89e051' },
  html: { glyph: '', color: '#e34c26' },
  css: { glyph: '', color: '#563d7c' },
  scss: { glyph: '', color: '#f55385' },
  yml: { glyph: '', color: '#6d8086' },
  yaml: { glyph: '', color: '#6d8086' },
  toml: { glyph: '', color: '#6d8086' },
  java: { glyph: '', color: '#cc3e44' },
  cs: { glyph: '', color: '#596706' },
  go: { glyph: '', color: '#519aba' },
  rs: { glyph: '', color: '#dea584' },
  sql: { glyph: '', color: '#dad8d8' },
  png: { glyph: '', color: '#a074c4' },
  jpg: { glyph: '', color: '#a074c4' },
  jpeg: { glyph: '', color: '#a074c4' },
  gif: { glyph: '', color: '#a074c4' },
  svg: { glyph: '', color: '#ffb13b' },
  lock: { glyph: '', color: '#bbbbbb' },
  txt: { glyph: '' },
};

/** Exact filenames that deserve their own icon regardless of extension. */
const FILENAME_ICONS: Record<string, FileIcon> = {
  '.gitignore': { glyph: '', color: '#e24329' },
  '.gitattributes': { glyph: '', color: '#e24329' },
  '.gitmodules': { glyph: '', color: '#e24329' },
  'package.json': { glyph: '', color: '#e8274b' },
  'package-lock.json': { glyph: '', color: '#7a0d21' },
  dockerfile: { glyph: '', color: '#458ee6' },
  makefile: { glyph: '', color: '#6d8086' },
  'readme.md': { glyph: '', color: '#519aba' },
  '.envrc': { glyph: '', color: '#faf743' },
  '.env': { glyph: '', color: '#faf743' },
};

/**
 * Devicon for a repo-relative file path — matched on exact filename first,
 * then extension, then a generic file glyph.
 */
export function fileIcon(filePath: string): FileIcon {
  const fileColour = getTheme().palette.file;
  if (asciiGlyphs) return { glyph: '·', color: fileColour };
  const name = (filePath.split('/').pop() ?? filePath).toLowerCase();
  const byName = FILENAME_ICONS[name];
  if (byName) return byName;
  const ext = name.includes('.') ? name.slice(name.lastIndexOf('.') + 1) : '';
  const byExt = EXTENSION_ICONS[ext];
  if (byExt) return { glyph: byExt.glyph, color: byExt.color ?? fileColour };
  return { glyph: DEFAULT_FILE_GLYPH, color: fileColour };
}

/* ── File status ────────────────────────────────────────────────────────── */

const FILE_BADGES: Record<FileStatusLetter, string> = {
  A: 'A ',
  M: 'M ',
  S: 'S ',
  MS: 'MS',
  D: 'D ',
  R: 'R ',
  U: 'U ',
  C: 'C ',
};

/** Exactly 2 display columns — right-aligned like the VS Code SCM gutter. */
export function tuiFileBadge(status: FileStatusLetter): string {
  return FILE_BADGES[status];
}

/** Semantic colour for a status letter. */
export function statusColor(status: FileStatusLetter): string {
  const p = getTheme().palette;
  switch (status) {
    case 'A':
    case 'S':
      return p.added;
    case 'M':
    case 'MS':
      return p.modified;
    case 'D':
    case 'U':
      return p.deleted;
    case 'R':
    case 'C':
      return p.renamed;
    default:
      return p.file;
  }
}

/** Map a FileChange to the same letter vocabulary as tree `statusLetter`. */
export function statusLetterFromChange(change: FileChange): FileStatusLetter {
  // Conflict before MS — staged+unstaged both set must not swallow U.
  if (change.unstagedStatus === 'U' || change.stagedStatus === 'U') return 'U';
  if (change.stagedStatus && change.unstagedStatus) return 'MS';
  const status = change.unstagedStatus ?? change.stagedStatus;
  if (status === 'R') return 'R';
  if (status === 'D') return 'D';
  if (change.untracked || status === 'A') return 'A';
  if (change.stagedStatus && !change.unstagedStatus) return 'S';
  if (status === 'C') return 'C';
  return 'M';
}

export function tuiFileBadgeForChange(change: FileChange): string {
  return tuiFileBadge(statusLetterFromChange(change));
}

/* ── Branch / sync ──────────────────────────────────────────────────────── */

/**
 * Merge-into-default mark for TUI branch chrome.
 * Maps plain `✅` / `🌱` to single-column glyphs (never emoji — Ink width).
 */
export function tuiMergeMark(merged: boolean | null): string {
  if (merged === true) return ICON_MERGED_INTO_DEFAULT;
  if (merged === false) return ICON_OPEN_VS_DEFAULT;
  return '';
}

/**
 * Sync mark: glyph plus commit count, e.g. ahead-by-2 or behind-by-3.
 */
export function tuiSyncMark(status: SyncStatus, note = ''): string {
  if (status === 'no-upstream') return ICON_NO_UPSTREAM;
  if (status === 'behind') {
    const count = note.match(/behind by (\d+)/)?.[1] ?? '';
    return `${ICON_BEHIND}${count}`;
  }
  if (status === 'ahead') {
    const count = note.match(/ahead by (\d+)/)?.[1] ?? '';
    return `${ICON_AHEAD}${count}`;
  }
  if (status === 'diverged') return ICON_DIVERGED;
  return ICON_SYNCED;
}

/** Colour matching `tuiSyncMark` semantics. */
export function syncColor(status: SyncStatus): string {
  const p = getTheme().palette;
  if (status === 'behind') return p.deleted;
  if (status === 'ahead') return p.added;
  if (status === 'diverged') return p.modified;
  return p.muted;
}

/**
 * Accent for {@link ICON_VIEWED}. Uses the theme cyan/blue token (`renamed`),
 * never the clean/success green (`added`).
 */
export function viewedColor(): string {
  return getTheme().palette.renamed;
}

/* ── Diff sections ──────────────────────────────────────────────────────── */

export type TuiSectionKind = 'staged' | 'unstaged' | 'new';

/** Diff pane section headers. */
export function tuiSectionHeader(kind: TuiSectionKind): string {
  if (kind === 'staged') return 'STAGED';
  if (kind === 'unstaged') return 'UNSTAGED';
  return 'NEW';
}

/** Accent colour for a diff section header. */
export function sectionColor(kind: TuiSectionKind): string {
  const p = getTheme().palette;
  if (kind === 'staged') return p.added;
  if (kind === 'new') return p.untracked;
  return p.modified;
}

/* ── Width helpers ──────────────────────────────────────────────────────── */

/** Truncate a string to at most `width` terminal columns (via `visibleWidth`). */
export function truncateVisible(value: string, width: number): string {
  if (width <= 0) return '';
  if (visibleWidth(value) <= width) return value;
  let out = '';
  let w = 0;
  for (const char of [...value]) {
    const cw = visibleWidth(char);
    if (w + cw > width) break;
    out += char;
    w += cw;
  }
  return out;
}

/**
 * Truncate a styled run to `width` columns, appending `…`.
 * Segments past the budget are dropped; the segment straddling the boundary is
 * cut mid-text so colour boundaries stay intact.
 */
export function truncateSegments(segments: Segment[], width: number): Segment[] {
  if (width <= 0) return [];
  let total = 0;
  for (const seg of segments) total += visibleWidth(seg.text);
  if (total <= width) return segments;

  const budget = Math.max(0, width - 1);
  const out: Segment[] = [];
  let used = 0;
  for (const seg of segments) {
    const segWidth = visibleWidth(seg.text);
    if (used + segWidth <= budget) {
      out.push(seg);
      used += segWidth;
      continue;
    }
    out.push({ ...seg, text: truncateVisible(seg.text, budget - used) });
    break;
  }
  out.push({ text: '…', dim: true });
  return out;
}

/** True if any codepoint is in the emoji / pictograph range (≥ U+1F300). */
export function hasWideEmoji(value: string): boolean {
  for (const char of [...value]) {
    const code = char.codePointAt(0) ?? 0;
    if (code >= 0x1f300) return true;
  }
  return false;
}

File: crates/workspace-status/src/tui/icons.rs
//! Nerd Font glyph registry for the ratatui tree. Port of Ink `src/tui/icons.ts`.
//!
//! A patched Nerd Font is a hard requirement. `WS_STATUS_GLYPHS=ascii` falls
//! back to plain markers. Every glyph occupies one terminal column: Nerd icons
//! live in the private-use area. Do not add emoji or CJK codepoints.

use crate::helpers::visible_width;
use crate::snapshot::{FileChange, SyncStatus};

/// Pick the Nerd Font glyph unless ASCII fallback is active.
pub fn glyph(ascii: bool, nerd: &'static str, fallback: &'static str) -> &'static str {
    if ascii {
        fallback
    } else {
        nerd
    }
}

/* ── Structure ──────────────────────────────────────────────────────────── */

/// Fold chevron — expanded. Width 1 in both modes (Ink does not ASCII-gate this).
pub const FOLD_EXPANDED: &str = "▾";
/// Fold chevron — collapsed.
pub const FOLD_COLLAPSED: &str = "▸";
/// ASCII fold expanded (Rust extra when `WS_STATUS_GLYPHS=ascii`).
pub const FOLD_EXPANDED_ASCII: &str = "v";
/// ASCII fold collapsed (Rust extra when `WS_STATUS_GLYPHS=ascii`).
pub const FOLD_COLLAPSED_ASCII: &str = ">";

/// Cursor accent bar painted in the left-most tree column.
pub const CURSOR_BAR: &str = "▌";

/// Vertical rule between panes and inside the diff gutter.
#[allow(dead_code)]
pub const RULE: &str = "│";

pub fn icon_workspace(ascii: bool) -> &'static str {
    glyph(ascii, "", "#")
}
pub fn icon_repo(ascii: bool) -> &'static str {
    glyph(ascii, "", "@")
}
/// Linked `git worktree` checkout. Nerd: nf-oct-link; ASCII: `L`.
pub fn icon_linked_worktree(ascii: bool) -> &'static str {
    glyph(ascii, "", "L")
}
pub fn icon_branch(ascii: bool) -> &'static str {
    glyph(ascii, "", "&")
}
/// Help MOVE column. Nerd: nf-dev-terminal_badge; ASCII: `+`.
pub fn icon_move(ascii: bool) -> &'static str {
    glyph(ascii, "", "+")
}
/// Help VIEW column. Nerd: nf-oct-diff; ASCII: `%`.
pub fn icon_diff(ascii: bool) -> &'static str {
    glyph(ascii, "", "%")
}
/// Font named in the Ink help footer (`REQUIRED_FONT`).
pub const REQUIRED_FONT: &str = "MesloLGM Nerd Font Mono";
pub fn icon_folder(ascii: bool) -> &'static str {
    glyph(ascii, "", "/")
}
#[allow(dead_code)]
pub fn icon_folder_open(ascii: bool) -> &'static str {
    glyph(ascii, "", "/")
}
pub fn icon_clean(ascii: bool) -> &'static str {
    glyph(ascii, "", ".")
}
pub fn icon_ignored(ascii: bool) -> &'static str {
    glyph(ascii, "", "~")
}
pub fn icon_ahead(ascii: bool) -> &'static str {
    glyph(ascii, "", "^")
}
pub fn icon_behind(ascii: bool) -> &'static str {
    glyph(ascii, "", "v")
}
pub fn icon_diverged(ascii: bool) -> &'static str {
    glyph(ascii, "", "Y")
}
pub fn icon_no_upstream(ascii: bool) -> &'static str {
    glyph(ascii, "", "?")
}
pub fn icon_synced(ascii: bool) -> &'static str {
    glyph(ascii, "", "=")
}
/// HEAD is an ancestor of the default-branch tip. Nerd: nf-fa-check-circle; ASCII: `M`.
pub fn icon_merged_into_default(ascii: bool) -> &'static str {
    glyph(ascii, "", "M")
}
/// HEAD is not merged into default. Nerd: nf-fa-tree; ASCII: `o`.
pub fn icon_open_vs_default(ascii: bool) -> &'static str {
    glyph(ascii, "", "o")
}
/// Ink `ICON_VIEWED` nerd glyph: nf-fa-eye (`U+F06E`).
///
/// Same pairing as `src/tui/icons.ts`. Do not substitute `◉` or another PUA eye.
pub const ICON_VIEWED_NERD: &str = "\u{f06e}";
/// Ink `ICON_VIEWED` ASCII fallback.
pub const ICON_VIEWED_ASCII: &str = "*";

/// Reviewed mark on a dirty file row. Nerd: nf-fa-eye (`U+F06E`); ASCII: `*`.
/// Distinct from `ICON_CLEAN` / `ICON_SYNCED`.
pub fn icon_viewed(ascii: bool) -> &'static str {
    glyph(ascii, ICON_VIEWED_NERD, ICON_VIEWED_ASCII)
}

const DEFAULT_FILE_GLYPH: &str = "";
const ASCII_FILE_GLYPH: &str = "·";

/// Devicon for a repo-relative file path. Exact filename wins, then extension.
pub fn file_icon(ascii: bool, file_path: &str) -> FileIcon {
    if ascii {
        return FileIcon {
            glyph: ASCII_FILE_GLYPH,
            color: None,
        };
    }
    let name = file_path
        .rsplit('/')
        .next()
        .unwrap_or(file_path)
        .to_ascii_lowercase();
    if let Some(icon) = filename_icon(&name) {
        return icon;
    }
    let ext = name.rsplit_once('.').map(|(_, ext)| ext).unwrap_or("");
    if let Some(icon) = extension_icon(ext) {
        return icon;
    }
    FileIcon {
        glyph: DEFAULT_FILE_GLYPH,
        color: None,
    }
}

/// Nerd file glyph plus optional hex colour (theme `file` when `None`).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct FileIcon {
    pub glyph: &'static str,
    pub color: Option<&'static str>,
}

fn filename_icon(name: &str) -> Option<FileIcon> {
    Some(match name {
        ".gitignore" | ".gitattributes" | ".gitmodules" => FileIcon {
            glyph: "",
            color: Some("#e24329"),
        },
        "package.json" => FileIcon {
            glyph: "",
            color: Some("#e8274b"),
        },
        "package-lock.json" => FileIcon {
            glyph: "",
            color: Some("#7a0d21"),
        },
        "dockerfile" => FileIcon {
            glyph: "",
            color: Some("#458ee6"),
        },
        "makefile" => FileIcon {
            glyph: "",
            color: Some("#6d8086"),
        },
        "readme.md" => FileIcon {
            glyph: "",
            color: Some("#519aba"),
        },
        ".envrc" | ".env" => FileIcon {
            glyph: "",
            color: Some("#faf743"),
        },
        _ => return None,
    })
}

fn extension_icon(ext: &str) -> Option<FileIcon> {
    Some(match ext {
        "ts" | "mts" | "cts" => FileIcon {
            glyph: "",
            color: Some("#519aba"),
        },
        "tsx" | "jsx" => FileIcon {
            glyph: "",
            color: Some("#519aba"),
        },
        "js" | "mjs" | "cjs" => FileIcon {
            glyph: "",
            color: Some("#cbcb41"),
        },
        "json" => FileIcon {
            glyph: "",
            color: Some("#cbcb41"),
        },
        "md" | "mdx" => FileIcon {
            glyph: "",
            color: Some("#519aba"),
        },
        "py" => FileIcon {
            glyph: "",
            color: Some("#ffbc03"),
        },
        "sh" | "bash" | "zsh" => FileIcon {
            glyph: "",
            color: Some("#89e051"),
        },
        "html" => FileIcon {
            glyph: "",
            color: Some("#e34c26"),
        },
        "css" => FileIcon {
            glyph: "",
            color: Some("#563d7c"),
        },
        "scss" => FileIcon {
            glyph: "",
            color: Some("#f55385"),
        },
        "yml" | "yaml" | "toml" => FileIcon {
            glyph: "",
            color: Some("#6d8086"),
        },
        "java" => FileIcon {
            glyph: "",
            color: Some("#cc3e44"),
        },
        "cs" => FileIcon {
            glyph: "",
            color: Some("#596706"),
        },
        "go" => FileIcon {
            glyph: "",
            color: Some("#519aba"),
        },
        "rs" => FileIcon {
            glyph: "",
            color: Some("#dea584"),
        },
        "sql" => FileIcon {
            glyph: "",
            color: Some("#dad8d8"),
        },
        "png" | "jpg" | "jpeg" | "gif" => FileIcon {
            glyph: "",
            color: Some("#a074c4"),
        },
        "svg" => FileIcon {
            glyph: "",
            color: Some("#ffb13b"),
        },
        "lock" => FileIcon {
            glyph: "",
            color: Some("#bbbbbb"),
        },
        "txt" => FileIcon {
            glyph: "",
            color: None,
        },
        _ => return None,
    })
}

/* ── File status ────────────────────────────────────────────────────────── */

/// Letter code aligned with Ink `FileStatusLetter`.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum FileStatusLetter {
    A,
    M,
    S,
    Ms,
    D,
    R,
    U,
    C,
}

impl FileStatusLetter {
    /// Ink letter token (`A` / `S` / `MS` / `M` / `D` / `R` / `U` / `C`).
    ///
    /// Watch signatures and tests use this, not the 2-column badge.
    pub fn as_str(self) -> &'static str {
        match self {
            Self::A => "A",
            Self::M => "M",
            Self::S => "S",
            Self::Ms => "MS",
            Self::D => "D",
            Self::R => "R",
            Self::U => "U",
            Self::C => "C",
        }
    }

    /// Exactly 2 display columns — right-aligned like the VS Code SCM gutter.
    pub fn badge(self) -> &'static str {
        match self {
            Self::A => "A ",
            Self::M => "M ",
            Self::S => "S ",
            Self::Ms => "MS",
            Self::D => "D ",
            Self::R => "R ",
            Self::U => "U ",
            Self::C => "C ",
        }
    }

    /// Ink `statusColor` token name.
    pub fn color_role(self) -> StatusColorRole {
        match self {
            Self::A | Self::S => StatusColorRole::Added,
            Self::M | Self::Ms => StatusColorRole::Modified,
            Self::D | Self::U => StatusColorRole::Deleted,
            Self::R | Self::C => StatusColorRole::Renamed,
        }
    }
}

/// Semantic colour for a status letter / sync mark.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum StatusColorRole {
    Added,
    Modified,
    Deleted,
    Renamed,
    Muted,
    #[allow(dead_code)]
    File,
}

/// Map a FileChange to the same letter vocabulary as Ink `statusLetterFromChange`.
pub fn status_letter_from_change(change: &FileChange) -> FileStatusLetter {
    // Conflict before MS — staged+unstaged both set must not swallow U.
    let unstaged = change.unstaged_status.as_deref();
    let staged = change.staged_status.as_deref();
    if unstaged == Some("U") || staged == Some("U") {
        return FileStatusLetter::U;
    }
    if staged.is_some() && unstaged.is_some() {
        return FileStatusLetter::Ms;
    }
    let status = unstaged.or(staged);
    if status == Some("R") {
        return FileStatusLetter::R;
    }
    if status == Some("D") {
        return FileStatusLetter::D;
    }
    if change.untracked || status == Some("A") {
        return FileStatusLetter::A;
    }
    if staged.is_some() && unstaged.is_none() {
        return FileStatusLetter::S;
    }
    if status == Some("C") {
        return FileStatusLetter::C;
    }
    FileStatusLetter::M
}

/// Exactly 2 display columns for a file-change badge.
pub fn tui_file_badge(change: &FileChange) -> &'static str {
    status_letter_from_change(change).badge()
}

/* ── Branch / sync ──────────────────────────────────────────────────────── */

/// Merge-into-default mark for TUI branch chrome. Never emoji.
pub fn tui_merge_mark(ascii: bool, merged: Option<bool>) -> &'static str {
    match merged {
        Some(true) => icon_merged_into_default(ascii),
        Some(false) => icon_open_vs_default(ascii),
        None => "",
    }
}

/// Sync mark: glyph plus commit count, e.g. ahead-by-2 or behind-by-3.
pub fn tui_sync_mark(ascii: bool, status: SyncStatus, note: &str) -> String {
    match status {
        SyncStatus::NoUpstream => icon_no_upstream(ascii).to_string(),
        SyncStatus::Behind => {
            let count = capture_count(note, "behind by ");
            format!("{}{count}", icon_behind(ascii))
        }
        SyncStatus::Ahead => {
            let count = capture_count(note, "ahead by ");
            format!("{}{count}", icon_ahead(ascii))
        }
        SyncStatus::Diverged => icon_diverged(ascii).to_string(),
        SyncStatus::UpToDate => icon_synced(ascii).to_string(),
    }
}

fn capture_count<'a>(note: &'a str, prefix: &str) -> &'a str {
    let Some(idx) = note.find(prefix) else {
        return "";
    };
    let start = idx + prefix.len();
    let end = note[start..]
        .find(|c: char| !c.is_ascii_digit())
        .map(|n| start + n)
        .unwrap_or(note.len());
    &note[start..end]
}

/// Colour matching `tui_sync_mark` semantics.
pub fn sync_color_role(status: SyncStatus) -> StatusColorRole {
    match status {
        SyncStatus::Behind => StatusColorRole::Deleted,
        SyncStatus::Ahead => StatusColorRole::Added,
        SyncStatus::Diverged => StatusColorRole::Modified,
        SyncStatus::UpToDate | SyncStatus::NoUpstream => StatusColorRole::Muted,
    }
}

/// Truncate a string to at most `width` terminal columns.
pub fn truncate_visible(value: &str, width: usize) -> String {
    if width == 0 {
        return String::new();
    }
    if visible_width(value) <= width {
        return value.to_string();
    }
    let mut out = String::new();
    let mut w = 0;
    for ch in value.chars() {
        let mut buf = [0u8; 4];
        let s = ch.encode_utf8(&mut buf);
        let cw = visible_width(s);
        if w + cw > width {
            break;
        }
        out.push(ch);
        w += cw;
    }
    out
}

/// True if any codepoint is in the emoji / pictograph range (≥ U+1F300).
#[allow(dead_code)]
pub fn has_wide_emoji(value: &str) -> bool {
    value.chars().any(|ch| (ch as u32) >= 0x1f300)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn change(untracked: bool, staged: Option<&str>, unstaged: Option<&str>) -> FileChange {
        FileChange {
            path: "a".into(),
            staged_status: staged.map(str::to_string),
            unstaged_status: unstaged.map(str::to_string),
            untracked,
            old_path: None,
        }
    }

    #[test]
    fn badges_are_two_columns_and_match_ink() {
        let letters = [
            FileStatusLetter::A,
            FileStatusLetter::M,
            FileStatusLetter::S,
            FileStatusLetter::Ms,
            FileStatusLetter::D,
            FileStatusLetter::R,
            FileStatusLetter::U,
            FileStatusLetter::C,
        ];
        let expected = ["A ", "M ", "S ", "MS", "D ", "R ", "U ", "C "];
        for (letter, want) in letters.into_iter().zip(expected) {
            assert_eq!(letter.badge(), want);
            assert_eq!(visible_width(letter.badge()), 2);
        }
        assert_eq!(
            status_letter_from_change(&change(true, None, None)),
            FileStatusLetter::A
        );
        assert_eq!(
            status_letter_from_change(&change(false, Some("M"), Some("M"))),
            FileStatusLetter::Ms
        );
        assert_eq!(
            status_letter_from_change(&change(false, Some("M"), None)),
            FileStatusLetter::S
        );
        assert_eq!(
            status_letter_from_change(&change(false, None, Some("M"))),
            FileStatusLetter::M
        );
        assert_eq!(
            status_letter_from_change(&change(false, Some("U"), Some("M"))),
            FileStatusLetter::U
        );
        assert_eq!(tui_file_badge(&change(true, None, None)), "A ");
    }

    #[test]
    fn structure_glyphs_are_width_one() {
        let ascii = [
            icon_workspace(true),
            icon_repo(true),
            icon_linked_worktree(true),
            icon_branch(true),
            icon_folder(true),
            icon_clean(true),
            icon_ignored(true),
            icon_ahead(true),
            icon_behind(true),
            icon_diverged(true),
            icon_no_upstream(true),
            icon_synced(true),
            icon_merged_into_default(true),
            icon_open_vs_default(true),
            icon_viewed(true),
            ASCII_FILE_GLYPH,
        ];
        let nerd = [
            icon_workspace(false),
            icon_repo(false),
            icon_linked_worktree(false),
            icon_branch(false),
            icon_folder(false),
            icon_clean(false),
            icon_ignored(false),
            icon_ahead(false),
            icon_behind(false),
            icon_diverged(false),
            icon_no_upstream(false),
            icon_synced(false),
            icon_merged_into_default(false),
            icon_open_vs_default(false),
            icon_viewed(false),
            CURSOR_BAR,
            FOLD_EXPANDED,
            FOLD_COLLAPSED,
            RULE,
        ];
        for g in ascii.into_iter().chain(nerd) {
            assert_eq!(visible_width(g), 1, "{g:?}");
            assert!(!has_wide_emoji(g), "{g:?}");
        }
        assert_eq!(icon_workspace(true), "#");
        assert_eq!(icon_viewed(true), "*");
        assert_ne!(icon_viewed(false), icon_clean(false));
        assert_ne!(icon_viewed(false), icon_synced(false));
    }

    #[test]
    fn viewed_is_ink_nf_fa_eye_not_a_substitute() {
        assert_eq!(icon_viewed(false), "\u{f06e}");
        assert_eq!(icon_viewed(true), "*");
        assert_eq!(icon_viewed(false), ICON_VIEWED_NERD);
        assert_eq!(icon_viewed(true), ICON_VIEWED_ASCII);
        let nerd = icon_viewed(false).chars().next().expect("glyph");
        assert_eq!(u32::from(nerd), 0xf06e);
        assert_ne!(nerd, '\u{25c9}'); // ◉
        assert_ne!(nerd, '\u{f07a}'); // other PUA eye/search lookalike
        assert_eq!(visible_width(icon_viewed(false)), 1);
    }

    #[test]
    fn file_icons_and_ascii_fallback() {
        assert_eq!(file_icon(true, "a.ts").glyph, "·");
        assert_eq!(visible_width(file_icon(false, "a.ts").glyph), 1);
        assert_eq!(visible_width(file_icon(false, "a.wat").glyph), 1);
        assert_ne!(
            file_icon(false, "package.json").glyph,
            file_icon(false, "tsconfig.json").glyph
        );
        assert_eq!(
            file_icon(false, "README.md").glyph,
            file_icon(false, "readme.md").glyph
        );
    }

    #[test]
    fn sync_and_merge_marks() {
        assert_eq!(
            tui_sync_mark(false, SyncStatus::Behind, "behind by 3"),
            format!("{}3", icon_behind(false))
        );
        assert_eq!(
            tui_sync_mark(true, SyncStatus::Ahead, "ahead by 2"),
            "^2".to_string()
        );
        assert_eq!(
            tui_sync_mark(true, SyncStatus::NoUpstream, ""),
            "?".to_string()
        );
        assert_eq!(
            tui_sync_mark(true, SyncStatus::Diverged, ""),
            "Y".to_string()
        );
        assert_eq!(
            tui_sync_mark(true, SyncStatus::UpToDate, ""),
            "=".to_string()
        );
        assert_eq!(tui_merge_mark(true, Some(true)), "M");
        assert_eq!(tui_merge_mark(true, Some(false)), "o");
        assert_eq!(tui_merge_mark(true, None), "");
    }
}
